libraries/chain/apply_context.cpp

```diff
@@ -258,8 +259,8 @@ uint32_t apply_context::execute_sync_call(name receiver, uint64_t flags, std::sp
    const auto max_sync_call_data_size = control.get_global_properties().configuration.max_sync_call_data_size;
    EOS_ASSERT(data.size() <= max_sync_call_data_size, sync_call_call_data_exception,
               "sync call call data size must be less or equal to ${s} bytes", ("s", max_sync_call_data_size));
-   EOS_ASSERT((flags & 0xFFFFFFFFFFFFFFFE) == 0, sync_call_validate_exception,
-              "sync call's flags ${f} can only set bit 0", ("f", flags));
+   EOS_ASSERT(flags <= static_cast<uint64_t>(sync_call_flags::last), sync_call_validate_exception,  // all but last `sync_call_flags::last` bits must be 0s
+              "only ${last} least significant bits of sync call's flags (${flags}) can be set", ("last", static_cast<uint64_t>(sync_call_flags::last))("flags", flags));
```

Comment:
Using `sync_call_flags::last` here doesn't seem right. If last is 1<<3 then this message will say "only 8 least significant bits of sync call's flags (${flags}) can be set". That value should be 4 instead of 8. I think you want the `std::bit_width(sync_call_flags::last)`

Reply:
Thanks for noticing this!

https://github.com/AntelopeIO/spring/pull/1266/commits/64eb9285134a2134ba9cf4efe569a55c48185fb0

Reply:
I think the check `flags <= static_cast<uint64_t>(sync_call_flags::last) is wrong. For example if two bits out of two were set, `flags` would be `3` and `static_cast<uint64_t>(sync_call_flags::last)` would be 2.

Reply:
Thanks. I renamed `last` and added comments to prevent future confusions. Also added flags validation tests.